minijinja/src/value/object.rs

```diff
@@ -101,3 +88,230 @@ impl<T: Object> Object for std::sync::Arc<T> {
+        self.range
+            .next()
+            .map(|idx| self.seq.get_item(idx).unwrap_or(Value::UNDEFINED))
+    }
+
+    #[inline(always)]
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        self.range.size_hint()
+    }
+}
+
+impl<'a> DoubleEndedIterator for SeqObjectIter<'a> {
+    #[inline(always)]
+    fn next_back(&mut self) -> Option<Self::Item> {
+        self.range
+            .next_back()
+            .map(|idx| self.seq.get_item(idx).unwrap_or(Value::UNDEFINED))
+    }
+}
+
+impl<'a> ExactSizeIterator for SeqObjectIter<'a> {}
+
+/// Views an [`Object`] as a struct.
+///
+/// # Example
+///
+/// The following example shows how to implement a dynamic object which
+/// represents a struct:
+///
+/// ```
+/// use std::fmt;
+/// use minijinja::value::{Value, Object, ObjectKind, StructObject};
+///
+/// #[derive(Debug, Clone)]
+/// struct Point(f32, f32, f32);
+///
+/// impl fmt::Display for Point {
+///     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+///         write!(f, "({}, {}, {})", self.0, self.1, self.2)
+///     }
+/// }
+///
+/// impl Object for Point {
+///     fn kind(&self) -> ObjectKind<'_> {
+///         ObjectKind::Struct(self)
+///     }
+/// }
+///
+/// impl StructObject for Point {
+///     fn get_field(&self, name: &str) -> Option<Value> {
+///         match name {
+///             "x" => Some(Value::from(self.0)),
+///             "y" => Some(Value::from(self.1)),
+///             "z" => Some(Value::from(self.2)),
+///             _ => None,
+///         }
+///     }
+///
+///     fn fields(&self) -> Box<dyn Iterator<Item = &str> + '_> {
+///         Box::new(["x", "y", "z"].into_iter())
+///     }
+/// }
+///
+/// let value = Value::from_object(Point(1.0, 2.5, 3.0));
+/// ```
+pub trait StructObject {
+    /// Invoked by the engine to get a field of a struct.
+    ///
+    /// Where possible it's a good idea for this to align with the return value
+    /// of [`fields`](Self::fields) but it's not necessary.
+    ///
+    /// If an field does not exist, `None` shall be returned.
+    ///
+    /// A note should be made here on side effects: unlike calling objects or
+    /// calling methods on objects, accessing fields is not supposed to
+    /// have side effects.  Neither does this API get access to the interpreter
+    /// [`State`] nor is there a channel to send out failures as only an option
+    /// can be returned.  If you do plan on doing something in field access
+    /// that is fallible, instead use a method call.
+    fn get_field(&self, idx: &str) -> Option<Value>;
+
+    /// Iterates over the fields.
+    ///
+    /// The default implementation returns an empty iterator.
+    fn fields(&self) -> Box<dyn Iterator<Item = &str> + '_> {
+        Box::new(None.into_iter())
+    }
+
+    /// Returns the number of fields in the struct.
+    ///
+    /// The default implementation returns the number of fields.
+    fn struct_size(&self) -> usize {
```

Comment:
And similarly here, `num_fields()` feels better to me.

Reply:
I will align them.